consensus/scheme/rolldpos/proposalpool.go

```diff
@@ -67,82 +67,47 @@ func (d *proposalPool) ReceiveBlock(blk *block.Block) error {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
-	// find the fork that contains this block
-	fork, err := d.forkAt(blk)
-	if err != nil {
-		return err
+	contain := func(node, leaf hash.Hash256) (exist bool) {
+		for b := leaf; ; {
+			if b == node {
+				return true
+			}
+			blk := d.nodes[b]
+			if blk == nil {
+				return false
+			}
+			b = blk.PrevHash()
+		}
 	}
 
 	// remove blocks in other forks or older blocks in the same fork
-	for f := range d.leaves {
-		start := d.nodes[f]
-		if f == fork {
+	leavesToDelete := make([]hash.Hash256, 0)
+	for leaf := range d.leaves {
+		start := d.nodes[leaf]
+		has := contain(blk.HashBlock(), leaf)
+		if has {
 			start = blk
 		}
 		for b := start; b != nil; b = d.nodes[b.PrevHash()] {
 			ha := b.HashBlock()
 			log.L().Debug("remove block from draft pool", log.Hex("hash", ha[:]), zap.Uint64("height", b.Height()), zap.Time("timestamp", b.Timestamp()))
 			delete(d.nodes, b.HashBlock())
 		}
+		if !has || blk.HashBlock() == leaf {
+			leavesToDelete = append(leavesToDelete, leaf)
```

Comment:
so just to confirm, for the case we discussed, this will delete A, B, 3, 4, and 1, 2 remain as 2 leaves?